### ISSUE TITLE
GPU memory use causes crash?

### ISSUE THREAD

**Issue by bensonbits** [sulcgroup/oxdna-web#22]
Hi guys, 

A colleague of mine tried to submit a DNA origami trimer (45k nt), and it caused oxDNA.org to "crash" after trying to use over 16 gb of GPU memory, log : 

INFO: pid: 28395
DEBUG: Debye-Huckel parameters: Q=0.054300, lambda_0=0.361645, lambda=0.357493, r_high=1.072479, cutoff=2.571515
DEBUG: Debye-Huckel parameters: debye_huckel_RC=1.608718e+00, debye_huckel_B=8.766183e-03
INFO: The Debye length at this temperature and salt concentration is 0.357493
INFO: (Cells.cpp) N_cells_side: 45, 45, 45; rcut=3.57152, IS_MC: 0
INFO: N: 43524
INFO: Using randomly distributed velocities
INFO: Initial kinetic energy: 12711.540039
INFO: Choosing device automatically
INFO: Computer has 8 devices
INFO:  - Trying device 0
INFO:  -- device 0 has properties 7.5
INFO:  -- using device 0
INFO:  --- Running on device 0
INFO: threads_per_block was not specified or set to 0. The default value (64) will be used
CUDA error at /opt/oxdna/oxDNA/src/CUDA/Lists/CUDASimpleVerletList.cu:88 code=2(cudaErrorMemoryAllocation) "GpuUtils::LR_cudaMalloc(&_d_cells, (size_t) _N_cells*_max_N_per_cell*sizeof(int))" 
INFO: CUDA max_neigh: 1340, max_N_per_cell: 43524, N_cells: 91125 (per side: 45 45 45)
INFO: CUDA Cells mem: 15129.91 MBs, lists mem: 222.65 MBs
INFO: Allocated CUDA memory: 16251.62 MBs
CUDA error at /opt/oxdna/oxDNA/src/CUDA/Lists/CUDASimpleVerletList.cu:88 code=2(cudaErrorMemoryAllocation) "GpuUtils::LR_cudaMalloc(&_d_cells, (size_t) _N_cells*_max_N_per_cell*sizeof(int))" 
/opt/oxdna/oxDNA/src/CUDA/Lists/CUDASimpleVerletList.cu(149) : getLastCudaError() CUDA error : fill_cells (SimpleVerlet) error : (2) out of memory.

When i run the same simulation on my computer it works fine and uses 2 gb of GPU memory,

INFO: pid: 7349
DEBUG: Debye-Huckel parameters: Q=0.054300, lambda_0=0.361645, lambda=0.505571, r_high=1.516714, cutoff=3.237868
DEBUG: Debye-Huckel parameters: debye_huckel_RC=2.275070e+00, debye_huckel_B=3.099314e-03
INFO: The Debye length at this temperature and salt concentration is 0.505571
INFO: (Cells.cpp) N_cells_side: 45, 45, 45; rcut=3.33787, IS_MC: 0
INFO: N: 43524
INFO: Using randomly distributed velocities
INFO: Initial kinetic energy: 12771.680664
INFO: Choosing device automatically
INFO: Computer has 1 devices
INFO:  - Trying device 0
INFO:  -- device 0 has properties 7.5
INFO:  -- using device 0
INFO:  --- Running on device 0
INFO: threads_per_block was not specified or set to 0. The default value (64) will be used
INFO: CUDA max_neigh: 1340, max_N_per_cell: 6858, N_cells: 91125 (per side: 45 45 45)
INFO: CUDA Cells mem: 2384.29 MBs, lists mem: 222.65 MBs
INFO: Allocated CUDA memory: 2614.57 MBs
INFO: SVN CODE VERSION: 6986

I can't really figure out what is causing this, is it some setting on oxdna.org ?

Cheers

Erik

**Comment by sulcgroup** [comment]
Hi Erik,
 The cards have 11GB of memory, so it tries to allocate more memory than is available. However, 45k nucleotides should still very much doable and we simulated much larger systems. It would help us if you are able to share the topology and the configuration to be able to find the issue. From the error message, I presume that you box in the configuration file is too large, so there is a lot of "empty space". The code however still allovates cell lists to keep track of nucleotides in all cells, causing the memory error.  I suggest making the box smaller (and making sure you do not break anything through periodic boundary condition) and then resubmitting.

**Comment by bensonbits** [comment]
Hi,

Thanks for getting back so quickly!

Here are the files, I had to manually shrink the box a bit (compared to what oxview suggested) to simulate  it on my computer and that works fine, but it seems like oxDNA.org automatically increases the box size again, making it so large that the simulation doesn't work? 

[Archive.zip](https://github.com/sulcgroup/oxdna-web/files/5828996/Archive.zip)

Cheers 

Erik

**Comment by bensonbits** [comment]
Just to clarify: the file that is tried to simulate has a box of 500 x 500 x 500, but when i open the crashed simulation oxdna.org (view inital config) the box looks much larger, so it seems like there is some automatic re-sizing of the box that is causing issues.

**Comment by ErikPoppleton** [comment]
@bensonbits 

oxView does change the size of the size of the box to 6*(max coordinate in any axis).  It will be a larger change to add an option that doesn't change the box size, but for now, I set the factor to 3 and I was able to start a simulation with your structure (box size = 650x650x650).  @sulcgroup I double checked that the example will still run. At 3* its possible for somebody to have a structure diagonally in an already improperly-sized box that will be able to interact through the PBC, however that would be a very specific edge case.
